### PR TITLE
pre-commit: test PR94648

### PR DESCRIPTION
**Pull request by dtcxzyw**
PR Link: https://github.com/llvm/llvm-project/pull/94648

### CODE REVIEW

File: bench/openssl/optimized/libcrypto-lib-ecp_oct.ll [hunk 1]
```diff
@@ -528,10 +528,11 @@ if.end:                                           ; preds = %entry
...
   %and1 = and i32 %conv, 254
-  %or.cond = icmp ugt i8 %0, 3
-  %1 = and i32 %conv, 252
-  %2 = icmp ne i32 %1, 4
-  %or.cond2 = and i1 %or.cond, %2
+  %1 = and i32 %conv, 2
+  %or.cond = icmp ne i32 %1, %and1
+  %2 = and i32 %conv, 252
+  %3 = icmp ne i32 %2, 4
+  %or.cond2 = and i1 %or.cond, %3
```
Comment (dtcxzyw) [inline bench/openssl/optimized/libcrypto-lib-ecp_oct.ll:535]: Regression.

File: bench/php/optimized/main.ll [hunk 2]
```diff
@@ -4390,18 +4390,22 @@ thread-pre-split37:                               ; preds = %18
...
-
-36:                                               ; preds = %.thread, %29, %24, %18, %12, %6, %1
+  %34 = trunc i64 %33 to i32
+  %35 = and i32 %34, 255
+  %36 = and i32 %34, 1
+  %or.cond = icmp ne i32 %36, %35
+  %37 = icmp ne i32 %35, 2
+  %or.cond5 = and i1 %or.cond, %37
+  %38 = trunc i64 %33 to i8
+  %spec.select = select i1 %or.cond5, i8 1, i8 %38
```
Comment (dtcxzyw) [inline bench/php/optimized/main.ll:4405]: Regression.

File: bench/postgres/optimized/mxactdesc.ll [hunk 3]
```diff
@@ -31,74 +31,75 @@ define dso_local void @multixact_desc(ptr noundef %0, ptr nocapture noundef read
   %6 = load ptr, ptr %5, align 8
   %7 = getelementptr inbounds i8, ptr %4, i64 56
   %8 = load i8, ptr %7, align 8
-  %or.cond = icmp ult i8 %8, 32
-  br i1 %or.cond, label %9, label %10
+  %9 = and i8 %8, -16
+  %10 = and i8 %8, 16
+  %or.cond = icmp eq i8 %10, %9
```
Comment (dtcxzyw) [inline bench/postgres/optimized/mxactdesc.ll:36]: Regression.

File: bench/qemu/optimized/hw_usb_dev-smartcard-reader.c.ll [hunk 4]
```diff
@@ -1340,8 +1340,9 @@ sw.bb96.i:                                        ; preds = %do.end74.i
   %bProtocolNum.i.i = getelementptr inbounds i8, ptr %call.i, i64 9175
   %19 = load i8, ptr %bProtocolNum.i.i, align 1
   %20 = and i8 %19, 3
-  %or.cond.i.i = icmp ugt i8 %20, 1
-  br i1 %or.cond.i.i, label %if.then.i.i, label %if.end.i.i
+  %21 = and i8 %19, 1
+  %or.cond.not.i.i = icmp eq i8 %21, %20
```
Comment (dtcxzyw) [inline bench/qemu/optimized/hw_usb_dev-smartcard-reader.c.ll:1344]: Regression.

File: bench/qemu/optimized/hw_nvme_ctrl.c.ll [hunk 5]
```diff
@@ -15663,30 +15663,31 @@ entry:
...
   br i1 %tobool.not, label %if.end, label %return
 
 if.end:                                           ; preds = %entry
-  %and = and i32 %.fr112, 255
-  %or.cond = icmp ugt i32 %and, 1
-  br i1 %or.cond, label %return, label %if.end8
+  %and = and i32 %.fr111, 255
+  %3 = and i32 %.fr111, 1
+  %or.cond.not = icmp eq i32 %3, %and
+  br i1 %or.cond.not, label %if.end8, label %return
```
Comment (dtcxzyw) [inline bench/qemu/optimized/hw_nvme_ctrl.c.ll:15677]: Regression.

File: bench/postgres/optimized/mxactdesc.ll (hunk 3 above)
Comment (goldsteinn) [reply bench/postgres/optimized/mxactdesc.ll:36]: This is weird, we handle:
```
define i1 @foo(i8 %x) {
  %x16 = and i8 %x, 16
  %xm16 = and i8 %x, -16
  %r = icmp eq i8 %x16, %xm16
  ret i1 %r
}
```

as:
```
define i1 @foo(i8 %x) {
  %r = icmp ult i8 %x, 32
  ret i1 %r
}
```

directly. Not sure why this ending up unsimplified. @nikic any ideas?

Comment (nikic) [reply bench/postgres/optimized/mxactdesc.ll:36]: Looks like a multi-use problem: https://llvm.godbolt.org/z/Y9zbMdqYK

Comment (nikic) [reply bench/postgres/optimized/mxactdesc.ll:36]: Note that there's an intermediate transform to:
```llvm

define i1 @foo(i8 %x) {
  %1 = and i8 -32, %x
  %r = icmp eq i8 %1, 0
  ret i1 %r
}
```
before we get to the single icmp form.

Comment (goldsteinn) [reply bench/postgres/optimized/mxactdesc.ll]: +1, I'll fix this up.

File: bench/icu/optimized/ucnv2022.ll [hunk 6]
```diff
@@ -138,50 +138,52 @@ if.then25:                                        ; preds = %if.then23
   br label %if.end219
 
 if.end26:                                         ; preds = %if.then23
-  %tobool30.not = icmp ult i32 %and, 2
-  br i1 %tobool30.not, label %if.end34, label %if.then63
+  %6 = and i32 %3, 1
+  %tobool30.not = icmp eq i32 %6, %and
+  br i1 %tobool30.not, label %if.end34, label %if.then31
```
Comment (goldsteinn) [inline bench/icu/optimized/ucnv2022.ll:143]: Maybe regression? I'm having a bit of trouble telling as the rest of the codegen looks slightly better.

**Pull request closed without merging**